pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ConsumerHandler.java

```diff
@@ -240,12 +242,27 @@ public void onWebSocketText(String message) {
                 // We should have received an ack
                 MessageId msgId = MessageId.fromByteArrayWithTopic(Base64.getDecoder().decode(command.messageId),
                         topic.toString());
-                consumer.acknowledgeAsync(msgId).thenAccept(consumer -> numMsgsAcked.increment());
-                if (!this.pullMode) {
-                    int pending = pendingMessages.getAndDecrement();
-                    if (pending >= maxPendingMessages) {
-                        // Resume delivery
-                        receiveMessage();
+                if ("cumulative".equals(command.ackType)) {
+                    ConsumerStats cStats = consumer.getStats();
+                    long ackDiff = cStats.getTotalMsgsReceived() - cStats.getTotalAcksSent();
+                    consumer.acknowledgeCumulativeAsync(msgId).thenAccept(consumer -> numMsgsAcked.add(ackDiff));
+                    if (!this.pullMode) {
+                        int acked = (int) ackDiff;
+                        IntBinaryOperator decrementFunction = (x, y) -> x - y;
+                        int pending = pendingMessages.accumulateAndGet(acked, decrementFunction);
```

Comment:
It is suitable to use `getAndAccumulate()` instead of ` accumulateAndGet()` here.

We must execute `receiveMessage()` in order to resume delivery of messages to the client only if `pendingMessages` *before* the update is greater than or equal to `maxPendingMessages`.

If `pendingMessages` is less than `maxPendingMessages`, then `receiveMessage()` has already been called just after receiving a message from the broker, and we don't need to execute it again here.
https://github.com/apache/pulsar/blob/918dc6842d09bc07478adb5fb099aa43f1d40f1b/pulsar-websocket/src/main/java/org/apache/pulsar/websocket/ConsumerHandler.java#L196-L200

Reply:
That's reasonable and easy to correct; thanks for the insight!